test/fake_cipherprov.c

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * https://www.openssl.org/source/license.html
+ * or in the file LICENSE in the source distribution.
+ */
+
+#include <string.h>
+#include <openssl/core_names.h>
+#include <openssl/core_object.h>
+#include <openssl/rand.h>
+#include <openssl/provider.h>
+#include <openssl/proverr.h>
+#include <openssl/param_build.h>
+#include "testutil.h"
+#include "fake_cipherprov.h"
+
+#define MAX_KEYNAME 32
+#define FAKE_KEY_LEN 16
+
+typedef struct prov_cipher_fake_ctx_st {
+    char key_name[MAX_KEYNAME];
+    unsigned char key[FAKE_KEY_LEN];
+} PROV_CIPHER_FAKE_CTX;
+
+static int ctx_from_key_params(PROV_CIPHER_FAKE_CTX *pctx, const OSSL_PARAM *params)
+{
+    const OSSL_PARAM *p;
+    char key_name[MAX_KEYNAME];
+    char *pval = key_name;
+    const void *val;
+    size_t used_len;
+
+    memset(key_name, 0, MAX_KEYNAME);
+
+    p = OSSL_PARAM_locate_const(params, FAKE_CIPHER_PARAM_KEY_NAME);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_utf8_string(p, &pval, MAX_KEYNAME)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+
+        memcpy(pctx->key_name, key_name, MAX_KEYNAME);
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_RAW_BYTES);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_octet_ptr(p, &val, &used_len)) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+
+        if (used_len > FAKE_KEY_LEN)
+            used_len = FAKE_KEY_LEN;
+
+        memcpy(pctx->key, val, used_len);
+    }
+
+    return 1;
+}
+
+static void fake_skeymgmt_free(void *keydata)
+{
+    OPENSSL_free(keydata);
+}
+
+static void *fake_skeymgmt_import(void *provctx, int selection, const OSSL_PARAM *p)
+{
+    PROV_CIPHER_FAKE_CTX *ctx = NULL;
+
+    if (!TEST_ptr(ctx = OPENSSL_zalloc(sizeof(PROV_CIPHER_FAKE_CTX))))
+        return 0;
+
+    if (ctx_from_key_params(ctx, p) != 1) {
+        OPENSSL_free(ctx);
+        return NULL;
+    }
+
+    return ctx;
+}
+
+static int fake_skeymgmt_export(void *keydata, int selection,
+                                OSSL_CALLBACK *param_callback, void *cbarg)
+{
+    OSSL_PARAM *params = NULL;
+    PROV_CIPHER_FAKE_CTX *ctx = (PROV_CIPHER_FAKE_CTX *)keydata;
+    int ret = 0;
+    OSSL_PARAM_BLD *tmpl = NULL;
+
+    tmpl = OSSL_PARAM_BLD_new();
```

Comment:
No need for the PARAM_BLD and this should use an OCTET_STRING instead of an OCTET_PTR.